src/Validation.hs

```diff
@@ -0,0 +1,395 @@
+6. 'Alternative': return first 'Success' or accumulate all errors inside 'Failure'.
+-}
+
+{-# LANGUAGE CPP                  #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Validation
+       ( -- * How to use
+         -- $use
+         Validation(..)
+       , validationToEither
+       , eitherToValidation
+       ) where
+
+import Control.Applicative (Alternative (..), Applicative (..))
+import Data.Bifunctor (Bifunctor (..))
+import Data.Foldable (Foldable (..))
+import Data.Kind (Constraint)
+import GHC.TypeLits (ErrorMessage (..), TypeError)
+
+#if MIN_VERSION_base(4,10,0)
+import Data.Bifoldable (Bifoldable (..))
+import Data.Bitraversable (Bitraversable (..))
+#endif
+
+-- >>> $setup
+-- >>> import Relude
+
+{- $use
+
+Take for example a type @Computer@ that needs to be validated:
+
+>>> :{
+data Computer = Computer
+    { computerRam  :: !Int  -- ^ Ram in Gigabytes
+    , computerCpus :: !Int
+    } deriving (Eq, Show)
+:}
+
+You can validate that the computer has a minimum of 16GB of RAM:
+
+>>> :{
+validateRam :: Int -> Validation [Text] Int
+validateRam ram
+    | ram >= 16 = Success ram
+    | otherwise = Failure ["Not enough RAM"]
+:}
+
+and that the processor has at least two CPUs:
+
+>>> :{
+validateCpus :: Int -> Validation [Text] Int
+validateCpus cpus
+    | cpus >= 2 = Success cpus
+    | otherwise = Failure ["Not enough CPUs"]
+:}
+
+You can use these functions with the 'Applicative' instance of the 'Validation'
+type to construct a validated @Computer@. You will get either (pun intended) a
+valid @Computer@ or the errors that prevent it from being considered valid.
+
+Like so:
+
+>>> :{
+mkComputer :: Int -> Int -> Validation [Text] Computer
+mkComputer ram cpus = Computer
+    <$> validateRam ram
+    <*> validateCpus cpus
+:}
+
+Using @mkComputer@ we get a @Success Computer@ or a list with all possible errors:
+
+>>> mkComputer 16 2
+Success (Computer {computerRam = 16, computerCpus = 2})
+
+>>> mkComputer 16 1
+Failure ["Not enough CPUs"]
+
+>>> mkComputer 15 2
+Failure ["Not enough RAM"]
+
+>>> mkComputer 15 1
+Failure ["Not enough RAM","Not enough CPUs"]
+-}
+
+-- | 'Validation' is 'Either' with a 'Left' that is a 'Semigroup'.
+data Validation e a
+    = Failure e
+    | Success a
+    deriving stock (Eq, Ord, Show)
+
+instance Functor (Validation e) where
+    fmap :: (a -> b) -> Validation e a -> Validation e b
+    fmap _ (Failure e) = Failure e
+    fmap f (Success a) = Success (f a)
+    {-# INLINE fmap #-}
+
+    (<$) :: a -> Validation e b -> Validation e a
+    x <$ Success _ = Success x
+    _ <$ Failure e = Failure e
+    {-# INLINE (<$) #-}
+
+{- | This instances covers the following cases:
+
+1. Both 'Success': combine values inside 'Success' with '<>'.
+2. Both 'Failure': combine values inside 'Failure' with '<>'.
+3. One 'Success', one 'Failure': return 'Failure'.
```

Comment:
This is a very low-level description of the code of this instance. Probably can add more high-level value of the instance?